File: app/src/main/java/rusha/x/BaseViewModel.java
package rusha.x;

import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.disposables.Disposable;

public abstract class BaseViewModel extends ViewModel {

    private List<Disposable> disposables = new ArrayList<>();

    protected void disposeOnExit(Disposable disposable) {
        disposables.add(disposable);
    }

    @Override
    public void onCleared() {
        super.onCleared();
        for (Disposable disposable : disposables) {
            disposable.dispose();
        }
    }
}

File: app/src/main/java/rusha/x/BaseViewModel.kt
package rusha.x

import androidx.lifecycle.ViewModel
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.Disposable

abstract class BaseViewModel : ViewModel() {

    private val disposables = mutableListOf<Disposable>()

    protected fun <T> Single<T>.managedSubscribe(onSuccess: (T) -> Unit): Disposable {
        val disposable = this.subscribe(onSuccess)
        disposables.add(disposable)
        return disposable
    }

    override fun onCleared() {
        super.onCleared()
        disposables.forEach { disposable ->
            disposable.dispose()
        }
    }
}